### ISSUE TITLE
cookbook don't catches the version change from normal attributes

### ISSUE THREAD

**Issue by timurb** [nathwill/chef-hekad#8]
Hi,

Looks like `default.ATTR.tap` doesn't work the way you intended it to work:
* when I set `heka.version` in node attributes to different version but `v0.9.2` it still downloads and installs 0.9.2

I've tried to `pp heka` just at the start of the recipe and here is what I get:
```ruby
{
  "release_url" =>" https://github.com/mozilla-services/heka/releases/download",
  "version" => "0.10.0b1",
  "tag" => "v0.9.2",
  "package"=> "heka_0.9.2_amd64.deb",
  "config"=> {"maxprocs"=>2}
}
```

I think just pulling parameters out from tap will solve the issue.

Do you need me to help you with fixing this?

**Comment by nathwill** [comment]
nope, that's just the way chef compiles attributes unfortunately. it loads this cookbook's attribute file before your attribute file, setting the hash keys before it comes back through to yours; yours then override the one attribute, but the others are untouched.

you can fix it by overriding all 3 in your cookbook, or we can fix it in this cookbook by getting rid of the package, tag, and version attributes in this cookbook and combining them all into the one release_url attribute, which is uglier, but less likely to confuse people... sound reasonable?

**Comment by timurb** [comment]
What I mean is if you write the attributes in the right part like:
```ruby
default['heka']['tag'] = "v#{node['heka']['version']}"
```
(instead of `heka['version']) this works ok: you _can_ override `heka.version` in node attributes. You just can't override it using `default` attribute in a wrapper cookbook.

I think `tap()` has some kind of caching implemented in it and that's why it works like that.

I've created a pull request https://github.com/nathwill/chef-hekad/pull/9 to fix that.

**Comment by nathwill** [comment]
iiinteresting... that sounds like an improvement if i understand correctly. i'll take a look at the PR and see what's up. Thanks!

**Comment by nathwill** [comment]
still not sure i understand how the PR is an improvement. i'm not seeing any different behavior (version 0.9.2 is still installed) when i try to apply this on top of your changes in order to test consuming this cookbook from a wrapper:

```diff
diff --git a/.kitchen.yml b/.kitchen.yml
index 1e16f1f..f6b92fb 100644
--- a/.kitchen.yml
+++ b/.kitchen.yml
@@ -16,4 +16,4 @@ platforms:
 suites:
   - name: hekad
     run_list:
-      - recipe[hekad]
+      - recipe[setup]
diff --git a/test/fixtures/cookbooks/setup/recipes/default.rb b/test/fixtures/cookbooks/setup/recipes/default.rb
index e69de29..933af78 100644
--- a/test/fixtures/cookbooks/setup/recipes/default.rb
+++ b/test/fixtures/cookbooks/setup/recipes/default.rb
@@ -0,0 +1,2 @@
+
+include_recipe 'hekad'
[nathwill@wyrd hekad]$ git add .
[nathwill@wyrd hekad]$ git diff --staged
diff --git a/.kitchen.yml b/.kitchen.yml
index 1e16f1f..f6b92fb 100644
--- a/.kitchen.yml
+++ b/.kitchen.yml
@@ -16,4 +16,4 @@ platforms:
 suites:
   - name: hekad
     run_list:
-      - recipe[hekad]
+      - recipe[setup]
diff --git a/test/fixtures/cookbooks/setup/attributes/default.rb b/test/fixtures/cookbooks/setup/attributes/default.rb
new file mode 100644
index 0000000..fe9539d
--- /dev/null
+++ b/test/fixtures/cookbooks/setup/attributes/default.rb
@@ -0,0 +1,2 @@
+
+override['heka']['version'] = '0.10.0b1'
diff --git a/test/fixtures/cookbooks/setup/recipes/default.rb b/test/fixtures/cookbooks/setup/recipes/default.rb
index e69de29..933af78 100644
--- a/test/fixtures/cookbooks/setup/recipes/default.rb
+++ b/test/fixtures/cookbooks/setup/recipes/default.rb
@@ -0,0 +1,2 @@
+
+include_recipe 'hekad'
```

**Comment by timurb** [comment]
Your patch didn't work for me either but this one works as I expected to:
```diff
diff --git a/.kitchen.yml b/.kitchen.yml
index 1e16f1f..a110b80 100644
--- a/.kitchen.yml
+++ b/.kitchen.yml
@@ -17,3 +17,6 @@ suites:
   - name: hekad
     run_list:
       - recipe[hekad]
+    attributes:
+      heka:
+        version: "0.10.0b1"
``` 

Not quite sure how Test Kitchen works here or probably Chef -- when reporting I've tested only node attributes and override attributes in wrapper cookbooks was only my guess.

**Comment by nathwill** [comment]
:/ bizarre... i bet that's setting the attribute *before* it comes through and parses the cookbook attributes file somehow? in any case, i went ahead and simplified this in https://github.com/nathwill/chef-hekad/pull/10, and shipped it with hekad v2.0.0.
